[ALLUXIO-2182] Some metrics improvements

peisun1115 opened the pull request:
1. Move the metrics lib to common so that we can use that outside of core/server
2. Make the metrics name unique including JvmSource. 
3. Add metrics to track #seeks over time.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11167/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11169/<h2>Failed Tests: <span class='status-failure'>2</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11169/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11169/org.alluxio$alluxio-tests/testReport/alluxio.master/AlluxioMasterRestApiTest/getMetricsTest/'><strong>alluxio.master.AlluxioMasterRestApiTest.getMetricsTest</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11169/org.alluxio$alluxio-tests/testReport/alluxio.worker/AlluxioWorkerRestApiTest/getMetricsTest/'><strong>alluxio.worker.AlluxioWorkerRestApiTest.getMetricsTest</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11170/<h2>Build result: FAILURE</span></h2>[...truncated 1861 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.3.0-SNAPSHOT/alluxio-shell-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.3.0-SNAPSHOT/alluxio-underfs-oss-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.3.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.3.0-SNAPSHOT/alluxio-underfs-oss-1.3.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.3.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.3.0-SNAPSHOT/alluxio-underfs-oss-1.3.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.3.0-SNAPSHOT/alluxio-keyvalue-client-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.3.0-SNAPSHOT/alluxio-underfs-hdfs-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.3.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.3.0-SNAPSHOT/alluxio-underfs-hdfs-1.3.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.3.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.3.0-SNAPSHOT/alluxio-underfs-hdfs-1.3.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.3.0-SNAPSHOT/alluxio-underfs-swift-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.3.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.3.0-SNAPSHOT/alluxio-underfs-swift-1.3.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.3.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.3.0-SNAPSHOT/alluxio-underfs-swift-1.3.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/pom.xml to org.alluxio/alluxio-underfs-s3a/1.3.0-SNAPSHOT/alluxio-underfs-s3a-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.3.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3a/1.3.0-SNAPSHOT/alluxio-underfs-s3a-1.3.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3a/target/alluxio-underfs-s3a-1.3.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3a/1.3.0-SNAPSHOT/alluxio-underfs-s3a-1.3.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.3.0-SNAPSHOT/alluxio-examples-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.3.0-SNAPSHOT/alluxio-core-server-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.3.0-SNAPSHOT/alluxio-keyvalue-common-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.3.0-SNAPSHOT/alluxio-underfs-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.3.0-SNAPSHOT/alluxio-underfs-glusterfs-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.3.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.3.0-SNAPSHOT/alluxio-underfs-glusterfs-1.3.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.3.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.3.0-SNAPSHOT/alluxio-underfs-glusterfs-1.3.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.3.0-SNAPSHOT/alluxio-integration-1.3.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.3.0-SNAPSHOT/alluxio-keyvalue-1.3.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11171/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11171/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11171/org.alluxio$alluxio-tests/testReport/alluxio.master/AlluxioMasterRestApiTest/getMetricsTest/'><strong>alluxio.master.AlluxioMasterRestApiTest.getMetricsTest</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11172/
Test PASSed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11176/
Test PASSed.

calvinjia commented on core/common/pom.xml:
```diff
@@ -78,6 +78,43 @@
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-server</artifactId>
+      <version>${jetty.version}</version>
+    </dependency>
```
I would like to avoid putting `jetty` into our client.

peisun1115 replied:
Done.

peisun1115 commented:
@calvinjia i've moved the MetricsServlet sink to core/server.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11199/
Test FAILed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11200/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-core-common' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11200/org.alluxio$alluxio-core-common/testReport'>org.alluxio:alluxio-core-common</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11200/org.alluxio$alluxio-core-common/testReport/alluxio.metrics/MetricsConfigTest/subPropertiesTest/'><strong>alluxio.metrics.MetricsConfigTest.subPropertiesTest</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11201/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-core-common' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11201/org.alluxio$alluxio-core-common/testReport'>org.alluxio:alluxio-core-common</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/11201/org.alluxio$alluxio-core-common/testReport/alluxio.metrics/MetricsConfigTest/subPropertiesTest/'><strong>alluxio.metrics.MetricsConfigTest.subPropertiesTest</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11202/
Test PASSed.

peisun1115 commented:
@calvinjia , do you have more comments?

calvinjia commented on core/common/src/main/java/alluxio/metrics/MetricsConfig.java:
```diff
@@ -144,11 +144,13 @@ private void parseConfiguration() {
   /**
    * Sets the default properties. The MetricsServlet is enabled and the path is /metrics/json
-   * by default.
+   * by default on master.
```
`on the servers`?

peisun1115 replied:
done

calvinjia commented on core/common/src/main/java/alluxio/metrics/MetricsSystem.java:
```diff
@@ -111,17 +115,29 @@ public MetricsSystem(String instance, MetricsConfig metricsConfig) {
+    if (instance.equals(MASTER_INSTANCE)) {
+      return Joiner.on(".").join(instance, source.getName());
+    } else {
+      return Joiner.on(".").join(instance, NetworkAddressUtils.getLocalHostName().replace('.', ','),
```
Is `,` the best replacement?

peisun1115 replied:
I have changed it to "_" now.

calvinjia commented on core/common/src/main/java/alluxio/metrics/MetricsSystem.java:
```diff
@@ -237,4 +254,23 @@ public void stop() {
+    String[] pieces = metricsName.split("\\.");
+    if (pieces.length <= 1) {
+      throw new IllegalArgumentException("Incorrect metrics name: " + metricsName);
+    }
```
Could this be replaced with a preconditions check?

peisun1115 replied:
done

calvinjia commented on core/common/src/main/java/alluxio/metrics/MetricsSystem.java:
```diff
@@ -237,4 +254,23 @@ public void stop() {
+   * @return the metrics name without instance and host name
+   */
+  public static String stripInstanceAndHost(String metricsName) {
+    String[] pieces = metricsName.split("\\.");
```
Instead of splitting and then joining later, would it be easier to follow if we tried pruning from the start of the string? (ie. check if master, prune `master.`, otherwise prune up to 2 `.`)

peisun1115 replied:
Oh, the current implementation is just simpler. Unless you are worried about perfomance, i would just keep this. 
The alternative you suggested will result in code like this.

id 1 = metricsName.indexOf(".")
substr = metricsName.substring(id1 + 1, ..)
id2 = substr.indexOf(".")
return substr.substring(...)

calvinjia commented on core/server/pom.xml:
```diff
@@ -42,53 +42,11 @@
-    </dependency>
-    <dependency>
-      <groupId>org.eclipse.jetty</groupId>
-      <artifactId>jetty-servlet</artifactId>
```
This dependency is no longer needed?

peisun1115 replied:
Added it back. It worked fine without this. I guess it is included by some other lib transitively

calvinjia replied:
Thanks, I think we try to explicitly import if we directly use something in the dependency.

calvinjia commented on core/common/src/main/java/alluxio/metrics/MetricsSystem.java:
```diff
@@ -111,17 +115,29 @@ public MetricsSystem(String instance, MetricsConfig metricsConfig) {
+  /**
+   * @return the MetricsServlet sink
+   */
+  public Sink getMetricsServlet() {
```
Should the naming be changed here?

peisun1115 replied:
hmm, what do you suggest? It is still MetricsServlet

calvinjia replied:
It feels like MetricsServlet shouldn't be understood by things in `common` since it is a concept in `servers` now. Is the motivation to name it this to avoid confusion when doing something like `(MetricsServlet) mc.getSink()`?

peisun1115 replied:
Agreed. I found a way to do it by iterating all the sinks.

calvinjia commented on core/common/src/main/java/alluxio/metrics/MetricsSystem.java:
```diff
@@ -53,7 +57,7 @@
   private MetricsConfig mMetricsConfig;
   private boolean mRunning = false;
-  private MetricsServlet mMetricsServlet;
+  private Sink mMetricsServlet;
```
It seems like the users of `getMetricsServlet` still want it as a `MetricsServlet`, would it be possible to move `MetricsServlet` into common and use that type?

peisun1115 replied:
then i will have to put the jetty lib in common which is not what you want?

calvinjia replied:
Oh I see, the return types are jetty objects. Nevermind

calvinjia commented on core/server/src/main/java/alluxio/worker/AlluxioWorkerRestServiceHandler.java:
```diff
@@ -233,7 +234,9 @@ public Response getMetrics() {
-        WorkerContext.getWorkerSource().getName(), WorkerSource.BLOCKS_CACHED);
+        MetricsSystem.buildSourceRegistryName(MetricsSystem.WORKER_INSTANCE,
+              WorkerContext.getWorkerSource()),
+        WorkerSource.BLOCKS_CACHED);
```
Can this be on one line?

peisun1115 replied:
done

calvinjia commented on tests/src/test/java/alluxio/master/AlluxioMasterRestApiTest.java:
```diff
@@ -113,7 +113,7 @@ public void getMetrics() throws Exception {
         new TypeReference<Map<String, Long>>() {});
 
-    Assert.assertEquals(Long.valueOf(0), metrics.get("master.CompleteFileOps"));
+    Assert.assertEquals(Long.valueOf(0), metrics.get("master.master.CompleteFileOps"));
```
Is this because the instance name is not stripped?

peisun1115 replied:
yeah.

calvinjia commented:
@peisun1115 I did a pass. What's the division between servers and common for the metrics system?

peisun1115 commented:
@calvinjia I have addressed your comments.

The division: common contains the general libraries to use the metrics system. It is basically a wrapper over the codehale metrics lib with some configuration logic.

The client/servers can register its own sources and special sinks such as the servlet sink.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11219/
Test PASSed.

calvinjia commented:
@peisun1115 I think my confusion comes from `MetricsServlet` being understood by `MetricSystem`. This is unavoidable if we want to keep jetty out of the client?

peisun1115 commented:
@calvinjia I have addressed your comments. PTAL. thanks

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11221/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11222/
Test PASSed.

calvinjia commented:
LGTM

peisun1115 commented:
@calvinjia can you merge it for me? thanks
